.circleci/config.yml

```diff
@@ -1,14 +1,111 @@
+# A note on performance.
+#
+# TL;DR: Storing and retrieving from workspaces is slow so use it sparingly.
+#
+# Workspaces are useful in sharing data between jobs where the overhead
+# involved in using a cache and providing functionality to rebuild the cache in
+# each job is too high.
+#
+# Long version:
+#
+# Storing and retrieving data from workspaces seems to be slow. And
+# caches/checkouts/plain downloads appear to be faster. Some numbers ...
+#
+####
+#
+# Restoring repo dir (no binaries) (65M) + gocache (243M) + gomodcache (1.5G)
+# from a workspace takes approx 121s
+#
+# Restoring repo dir (no binaries) (65M) + gocache (243M) from a workspace
+# takes approx 52s and then restoring the gomodcache (1.5G) from a cache takes
+# approx 12s, so 64s total, so a reduction of 57s which is almost half the
+# time!
+#
+####
+#
+# Restoring monorepo dir + celo-blockchain  + go install (684M) from workspace
+# approx 51s.
+#
+# Restoring monorepo dir + go install (659M) from workspace approx 37s +
+# shallow ceckout celo-monorepo 2s, so a total of 39s a 12 second improvement.
+#
+####
+#
+# Restoring monorepo dir + golang install from workspace (659M) takes
+# approx 35s.
+#
+# Downloading and installing go (109M) takes approx 4s and restoring monorepo
+# dir from (414M) takes approx 28s, so a total of 32s, so a 3s improvement.
+#
+#######
+#
+# As can be seen from the above numbers restoring from workspaces are
+# significantly slower than caches, a fair bit slower than github checkouts and
+# a little bit slower than downloading and installing go.
+#
+# Since workspaces are scoped to a workflow you need to persist to a workspace
+# in your workflow before you can read from it, which further slows the build
+# time and persisting seems to take about twice as long as retrieving. And on
+# top of that remember that before persisting you will need to run the task
+# that you would otherwise run directly in a downstream job.
+#
+# Workspaces work best for small amounts of data that take a long time to
+# generate that we want to share between many jobs without having to duplicate
+# everywhere the logic to generate the data.
+#
+# How are we using them?
+#
+# We are currently using workspaces to store the compiled-system-contracts
+# (4.7M) which take about 5 minutes to generate 2s to persist to the workspace
+# and 0s to restore from the workspace, we also cache them so that most of the
+# time we can skip the build step and simply persist the cache to the
+# workspace. Having this functionality contained in one job makes the config
+# simpler than if we tried to add compiled system contract generation to all of
+# the downstream jobs and since this job is quite fast and not in the critical
+# path for the workflow it's ok to be a bit slower here.
+#
+########
```

Comment:
I suggest putting this section `The structure of this workflow` at the top, making it easier to understand above performance note.